File: app/src/main/java/upday/droidconmvvm/MainActivity.java
package upday.droidconmvvm;

import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import upday.droidconmvvm.model.Language;

public class MainActivity extends AppCompatActivity {

    @NonNull
    private CompositeDisposable mCompositeDisposable;

    @NonNull
    private MainViewModel mViewModel;

    @Nullable
    private TextView mGreetingView;

    @Nullable
    private Spinner mLanguagesSpinner;

    @Nullable
    private LanguageSpinnerAdapter mLanguageSpinnerAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mViewModel = getViewModel();
        setupViews();
    }

    private void setupViews() {
        mGreetingView = (TextView) findViewById(R.id.greeting);

        mLanguagesSpinner = (Spinner) findViewById(R.id.languages);
        assert mLanguagesSpinner != null;
        mLanguagesSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(final AdapterView<?> parent, final View view,
                                       final int position, final long id) {
                itemSelected(position);
            }

            @Override
            public void onNothingSelected(final AdapterView<?> parent) {
                //nothing to do here
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        bind();
    }

    @Override
    protected void onPause() {
        super.onPause();
        unBind();
    }

    private void bind() {
        mCompositeDisposable = new CompositeDisposable();

        mCompositeDisposable.add(mViewModel.getGreeting()
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::setGreeting));

        mCompositeDisposable.add(mViewModel.getSupportedLanguages()
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::setLanguages));
    }

    private void unBind() {
        mCompositeDisposable.clear();
    }

    private void setGreeting(@NonNull final String greeting) {
        assert mGreetingView != null;

        mGreetingView.setText(greeting);
    }

    private void setLanguages(@NonNull final List<Language> languages) {
        assert mLanguagesSpinner != null;

        mLanguageSpinnerAdapter = new LanguageSpinnerAdapter(this,
                R.layout.language_item,
                languages);
        mLanguagesSpinner.setAdapter(mLanguageSpinnerAdapter);
    }

    @NonNull
    private MainViewModel getViewModel() {
        return ((DroidconApplication) getApplication()).getViewModel();
    }

    private void itemSelected(final int position) {
        assert mLanguageSpinnerAdapter != null;

        Language languageSelected = mLanguageSpinnerAdapter.getItem(position);
        mViewModel.languageSelected(languageSelected);
    }
}

File: app/src/main/java/upday/droidconmvvm/MainActivity.kt
package upday.droidconmvvm

import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import upday.droidconmvvm.databinding.ActivityMainBinding
import upday.droidconmvvm.model.Language

class MainActivity : AppCompatActivity() {

    private val mViewModel: MainViewModel by viewModels {
        MainViewModelFactory(
            (application as DroidconApplication).dataModel
        )
    }

    private lateinit var mLanguageSpinnerAdapter: LanguageSpinnerAdapter

    private lateinit var _activityView: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _activityView = ActivityMainBinding.inflate(layoutInflater)
        setContentView(_activityView.root)
        setupObservers()
    }

    private fun setupObservers() {

        setLanguages(mViewModel.supportedLanguages)

        val selectedLanguageObserver = Observer<String> { greeting ->
            setGreeting(greeting)
        }
        mViewModel.currentGreeting.observe(this, selectedLanguageObserver)

        _activityView.languagesSpinner.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>?, view: View?,
                position: Int, id: Long
            ) {
                itemSelected(position)
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {
                //nothing to do here
                Log.d(this.javaClass.name, "cayo aqui")
            }
        }
    }

    private fun setGreeting(greeting: String) {
        _activityView.greetingTextView.text = greeting
    }

    private fun setLanguages(languages: List<Language>) {
        mLanguageSpinnerAdapter = LanguageSpinnerAdapter(
            this,
            R.layout.language_item,
            languages
        )
        _activityView.languagesSpinner.adapter = mLanguageSpinnerAdapter
    }

    private fun itemSelected(position: Int) {
        val languageSelected = mLanguageSpinnerAdapter.getItem(position)
        mViewModel.setLanguageSelected(languageSelected!!)
    }
}